Revamp showcase

ineiti opened the pull request:
This is the main branch where all the different changes are merged little by little, before putting everything into main.

Closes #206 #207 #209 #210 #211 #214 #215 #216 #217 #218 #220 #221 #222 #223

tharvik commented:
* "Search:" to "Filter:"
* in search bar and dark mode, the text is light grey, which is hard to read on white
* having an already field search bar is weird
* you can't see which colum is selected in the columns list in dark mode
* on "clear search", the categories disappear
* when selecting "All project" & one category, nothing appears
* link to "labs" is dead
* when selecting a lab/artifact and clearing the search, it is not filtered but the filters are still showing the old values
* the bar below is not showing the correct columns; get rid of it even?
* maybe do not repeat the category in the list when filtering with one
* the greys for inactive projects are not the same as before, is it wanted?

tharvik commented on showcase.py:
```diff
@@ -21,6 +22,15 @@
 # Main Factory page
 FACTORY_URL = "https://www.c4dt.org/factory/software/"
 
+CATEGORIES = {
+    "Privacy": [1, "Privacy Protection & Cryptography"],
+    "Blockchain": [2, "Blockchains & Smart Contracts"],
+    "Verification": [3, "Software Verification"],
+    "Security": [4, "Device and System Security"],
+    "Learning": [5, "Machine Learning"],
+    "Other": [6, "Other"],
+}
```
Why not a `List[Tuple[str, str]]`?

tharvik commented on showcase.py:
```diff
@@ -50,6 +60,11 @@ def find_project(project_id, lab_id=None):
     bottle.abort(404, f"Project '{project_id}' does not exist")
 
+def find_project_tabs(project_id):
+    tabs = filter(lambda tab: os.path.isfile(os.path.join("views", "incubator", tab, project_id + ".tpl")),
+                  ["presentation", "background", "demo", "hands-on", "pilot", "app"])
+    return list(tabs) + ["technical"]
```
`filter` is deprecated, it's better to use a for comprehension

tharvik commented on test_app.py:
```diff
@@ -56,49 +55,39 @@ def test_consistent_data():
-            showcase.projects(lab_id)
 
 def test_all_projects():
     labs = data.load()
 
+    # with patch.object(showcase, 'find_project_tabs', return_value=['technical']):
```
commented code?

tharvik reviewed (commented):
* on the project pages, the black border of the breadcrumbs is cut
* on the project pages, there is a grey background for the content of each tab, is it wanted or would no-background be more readable?

ineiti commented:
> * on the project pages, the black border of the breadcrumbs is cut

On Chrome the border is there, just very much into the border of the window. I can add a margin to make the border more visible, but then the text behind the menu is visible again and needs an extra div to hide it. So I prefer keeping it like that.

> * on the project pages, there is a grey background for the content of each tab, is it wanted or would no-background be more readable?

That's on purpose. If David and Stephanie complain, too, I'll remove it.

cgrigis commented:
Generic comments (tested on Firefox):

* [ ] The categories ("blockchain", "privacy", etc.) disappear as soon as we sort a column, enter a search or select a tag.
* [ ] Filtering using a tag does not give a feedback on what the filter is, and it is not clear how to remove it.
* [ ] The "Clear search" button is no longer there.
* [ ] Zooming out (CTRL -), then zooming back in (CTRL +) does not bring back the same layout.
* [ ] Selecting a tag, then entering a search term seems to clear the tag filtering.
* [ ] Selecting some tags seems to bring the browser in an infinite loop ("A web page is slowing down your browser. What would you like to do?"). Consistently happening with "Privacy Preserving".
* [ ] If displayed, the "Active" and "Incubator" columns don't get row coloring (maybe they should not be displayable?).
* [ ] In the project views, the black line under the breadcrumbs is interrupted where the info id displayed.
* [ ] When a project is selected, the breadcrumbs no longer match the URL (i.e. "Incubator" vs "Showcase").
* [ ] The project name in the breadcrumbs does not follow the project case (i.e. displaying "Spindle" instead of "SPINDLE", "At2" instead of "AT2", or "Free gaen alternative" instead of "Free GAEN alternative").

cgrigis commented on views/incubator/demo/stainless.tpl:
```diff
@@ -0,0 +1,28 @@
+    could have prevented it.
+</p>
+
+<p>The following screenshot shows the start of the demonstrator. If you want
+    to discover for yourself, you can try to change the contract with (1). But a
+    better starting point is the tutorial, which you'll find in (2).</p>
```
I would not suggest to the user that they try to "discover", but rather that they stick to the tutorial at first. It's already confusing enough as it is ;-)

Something like:
```suggestion
<p>The following screenshot shows the start of the demonstrator. The tutorial
    will guide you through the various elements of the interface and help you
    on your way.</p>
```

ineiti commented on showcase.py:
```diff
@@ -50,6 +60,11 @@ def find_project(project_id, lab_id=None):
+def find_project_tabs(project_id):
+    tabs = [ tab for tab in ["presentation", "background", "demo", "hands-on", "pilot", "app"]
+             if os.path.isfile(os.path.join("views", "incubator", tab, project_id + ".tpl"))]
+    return list(tabs) + ["technical"]
```
yes, now it is. Before it was the result of a `filter`...

ineiti commented:
> * [x]  The categories ("blockchain", "privacy", etc.) disappear as soon as we sort a column, enter a search or select a tag.

That's on purpose, at least for the sorting. For the search the categories stay if they are targeted. Same thing for the tag.

> * [x]  Filtering using a tag does not give a feedback on what the filter is, and it is not clear how to remove it.

Added tags to the search box.

> * [x]  The "Clear search" button is no longer there.

Added a white X-button to the right

> * [ ]  Zooming out (CTRL -), then zooming back in (CTRL +) does not bring back the same layout.

Oh my - more CSS black magic.

> * [x]  Selecting a tag, then entering a search term seems to clear the tag filtering.

That is clearer now, as the tags show in the search box.

> * [ ]  Selecting some tags seems to bring the browser in an infinite loop ("A web page is slowing down your browser. What would you like to do?"). Consistently happening with "Privacy Preserving".

Works here - perhaps got magically fixed with the tags appearing in the search-box? Can you try again, please?

> * [ ]  If displayed, the "Active" and "Incubator" columns don't get row coloring (maybe they should not be displayable?).

I searched in the datables.net, but couldn't find how I can make them non-selectable. Any idea?

> * [x]  In the project views, the black line under the breadcrumbs is interrupted where the info id displayed.

Oh f*** - I thought these days are over. Presents perfect on Chrome... So if I use `top: 2.25em` instead of `top: 2.2em`, it shows correctly on both. Manual calculation shows it should be `top: 2.4em`, but then there is a gap and the text shows... Even more fun: Safari shows the text in the window-header.

> * [x]  When a project is selected, the breadcrumbs no longer match the URL (i.e. "Incubator" vs "Showcase").

Yes, that's what we discussed with Valérian. Before it was `Factory - Showcase - Spindle - Presentation`, and if you clicked on `Spindle`, it went back to Presentation. So we removed the tab-name at the end.

> * [x]  The project name in the breadcrumbs does not follow the project case (i.e. displaying "Spindle" instead of "SPINDLE", "At2" instead of "AT2", or "Free gaen alternative" instead of "Free GAEN alternative").

Oups - I thought I did a good thing using `capitalize`, but that would've been necessary for `project_id`. And I'm using `project['name']` anyway, so I removed `capitalize` and now it all shows nicely.



So I think I need some help for the open `[ ]`...

cgrigis commented:
> > * [x]   The categories ("blockchain", "privacy", etc.) disappear as soon as we sort a column, enter a search or select a tag.
> 
> That's on purpose, at least for the sorting. For the search the categories stay if they are targeted. Same thing for the tag.

Ok. I personally find it confusing, particularly because after I sort on a column, I don't know how to go back to the previous view (short of reloading the page, or selecting a different entry in the dropdown lists and going back to the previous entry).

> > * [ ]   Zooming out (CTRL -), then zooming back in (CTRL +) does not bring back the same layout.
> 
> Oh my - more CSS black magic.

Indeed :-/  

> > * [ ]   Selecting some tags seems to bring the browser in an infinite loop ("A web page is slowing down your browser. What would you like to do?"). Consistently happening with "Privacy Preserving".
> 
> Works here - perhaps got magically fixed with the tags appearing in the search-box? Can you try again, please?

It currently happens consistently to me. How to reproduce:
- Open Firefox
- Go to https://show-new.c4dt.org/showcase/
- Click on the very first tag for SPINDLE "Privacy Preserving"

> > * [ ]   If displayed, the "Active" and "Incubator" columns don't get row coloring (maybe they should not be displayable?).
> 
> I searched in the datables.net, but couldn't find how I can make them non-selectable. Any idea?

Not off the top of my head. If that doesn't work, extend the coloring to those columns?

> > * [x]   In the project views, the black line under the breadcrumbs is interrupted where the info id displayed.
> 
> Oh f*** - I thought these days are over. Presents perfect on Chrome... So if I use `top: 2.25em` instead of `top: 2.2em`, it shows correctly on both. Manual calculation shows it should be `top: 2.4em`, but then there is a gap and the text shows... Even more fun: Safari shows the text in the window-header.

:sob: 

> > * [x]   When a project is selected, the breadcrumbs no longer match the URL (i.e. "Incubator" vs "Showcase").
> 
> Yes, that's what we discussed with Valérian. Before it was `Factory - Showcase - Spindle - Presentation`, and if you clicked on `Spindle`, it went back to Presentation. So we removed the tab-name at the end.

Sorry, what I meant was that when we select a project, the Breadcrums shows e.g.  "Factory > Showcase > Spindle", but the URL is "https://show-new.c4dt.org/incubator/spindle/presentation" ("Incubator" vs "Showcase")

ineiti commented:
> Ok. I personally find it confusing, particularly because after I sort on a column, I don't know how to go back to the previous view (short of reloading the page, or selecting a different entry in the dropdown lists and going back to the previous entry).

Now when the search box is cleared it goes back to showing the titles

> > > * [ ]    Zooming out (CTRL -), then zooming back in (CTRL +) does not bring back the same layout.

Willnotfix. If you use zoom-in and zoom-out, you'll have to reload, sorry...

> * Open Firefox
> * Go to https://show-new.c4dt.org/showcase/
> * Click on the very first tag for SPINDLE "Privacy Preserving"

Not reproducible here.

> > > * [X]    If displayed, the "Active" and "Incubator" columns don't get row coloring (maybe they should not be displayable?).
> > > * [x]    In the project views, the black line under the breadcrumbs is interrupted where the info id displayed.

> > > * [x]    When a project is selected, the breadcrumbs no longer match the URL (i.e. "Incubator" vs "Showcase").
> Sorry, what I meant was that when we select a project, the Breadcrums shows e.g. "Factory > Showcase > Spindle", but the URL is "https://show-new.c4dt.org/incubator/spindle/presentation" ("Incubator" vs "Showcase")

Done

ineiti commented:
OK, last iteration, promised. Only taking code-comments, comments with regard to the structure itself go in new issues for later fixing.

cgrigis commented on data/TAGS.txt:
```diff
@@ -0,0 +1,146 @@
+- application verticals
+- technical aspects
+
+The tag-list here is sorted alphabetically, with all tags from one letter comma-separated.
```
?

ineiti replied:
That I didn't follow my spec ;) I wanted to have 26 lines, one for each letter, and then comma-separate the tags for each starting letter...

cgrigis commented on views/incubator/technical.tpl:
```diff
@@ -0,0 +1,161 @@
+        <%
+        active = is_active(project)
+        if active is None:
+        status = "unknown"
```
indentation?

cgrigis commented on views/incubator/technical.tpl:
```diff
@@ -0,0 +1,161 @@
+        if active is None:
+        status = "unknown"
+        else:
+        status = "active" if active else "inactive"
```
indentation?

cgrigis commented on data.py:
```diff
@@ -58,7 +74,6 @@
                 "email": sy.Email(),
                 }),
-            sy.Optional("c4dt_work"): sy.Str(),
             "tags": sy.Seq(sy.Str()),
```
Why not use an enumerate type for the tags, now that we want to enforce a fixed list? This would provide an automatic check for consistency.

ineiti replied:
I suppose the tag list is changing much more frequently than the category or any other list. So I'm hesitant to put an often changing list here. What do you think?

ineiti replied:
OK, added all the tags which showed some wrong ones...
